From 921e47ee1d9f0ecc73226a96e1ee4e7d7517f123 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Wed, 31 Jan 2024 11:41:46 +0100
Subject: [PATCH 525/841] feat(stm32mp2): manage RISAF MCE key for STM32MP21

On STM32MP21, RISAF encryption key is automatically generated from the
HWKEY (derivation) using RNG2 IP and enabled by a dummy write sequence
on RISAF encryption key registers.
The context seed saving can then be replaced by new MCE one.
Set maximum size at platform level.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 73fb2028de6d5a47189d0510d3a108ccf3cd6019 feat(stm32mp2): manage RISAF MCE key for STM32MP21]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I3f1571eaff1749758c7ed3d7522d629f8a7eb865
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/356006
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 6 ++++++
 plat/st/stm32mp2/stm32mp2_def.h   | 4 ++++
 2 files changed, 10 insertions(+)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index b34985e1e..198124576 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -503,6 +503,12 @@ static void prepare_encryption(void)
 	if (stm32mp2_risaf_write_encryption_key(RISAF4_INST, mkey) != 0) {
 		panic();
 	}
+
+#if STM32MP21
+	if (stm32mp2_risaf_write_mce_key(RISAF4_INST, mkey) != 0) {
+		panic();
+	}
+#endif /* STM32MP21 */
 }
 #endif /* STM32MP_M33_TDCID */
 
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 70da89b75..6042a7b32 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -620,7 +620,11 @@ enum ddr_type {
 #define RISAF_MAX_REGION			(RISAF1_MAX_REGION + RISAF2_MAX_REGION + \
 						 RISAF4_MAX_REGION + RISAF5_MAX_REGION)
 
+#if STM32MP21
+#define RISAF_KEY_SIZE_IN_BYTES			RISAF_MCE_KEY_256BITS_SIZE_IN_BYTES
+#else /* STM32MP21 */
 #define RISAF_KEY_SIZE_IN_BYTES			RISAF_ENCRYPTION_KEY_SIZE_IN_BYTES
+#endif /* STM32MP21 */
 #define RISAF_SEED_SIZE_IN_BYTES		U(4)
 
 /*******************************************************************************
-- 
2.47.2

